hedera-node/src/main/java/com/hedera/services/state/migration/MigrationRecordsManager.java

```diff
@@ -224,13 +273,101 @@ private void publishContractFreeAutoRenewalRecords() {
                         });
     }
 
-    @VisibleForTesting
-    void setSideEffectsFactory(Supplier<SideEffectsTracker> sideEffectsFactory) {
-        this.sideEffectsFactory = sideEffectsFactory;
+    private void publishTraceabilityMigrationRecords() {
+        final var contractStorageMap = contractStorage.get();
+        accounts.get()
+                .forEach(
+                        (id, account) -> {
+                            if (!account.isSmartContract()) {
+                                return;
+                            }
+                            final var contractId = id.toGrpcContractID();
+
+                            final var bytecodeSidecar =
+                                    generateMigrationBytecodeSidecarFor(contractId);
+                            transactionContext.addSidecarRecord(bytecodeSidecar);
+                            log.debug(
+                                    "Published migration bytecode sidecar for contract 0.0.{}",
+                                    contractId.getContractNum());
+
+                            var contractStorageKey = account.getFirstContractStorageKey();
+                            if (contractStorageKey == null) {
+                                log.debug(
+                                        "Contract 0.0.{} has no iterable storage - no migration"
+                                                + " state changes will be published.",
+                                        contractId.getContractNum());
+                                return;
+                            }
+                            final var stateChangesSidecar =
```

Comment:
Even though we have no reason to think the storage list _could_ have a loop, "Only the paranoid survive", as they say!

So I'd suggest passing in `account.getNumContractKvPairs()` as a limit on the maximum number of key/value pairs we iterate through.

Reply:
Added logic now to protect against such a scenario